octodns/manager.py

```diff
@@ -243,7 +243,8 @@ def _populate_and_plan(self, zone_name, sources, targets, desired=None,
                 try:
                     source.populate(zone, lenient=lenient)
                 except TypeError as e:
-                    if "keyword argument 'lenient'" not in text_type(e):
+                    if ("unexpected keyword argument 'lenient'"
+                            not in text_type(e)):
```

Comment:
Why was the extra word `unexpected` needed here for python 2.7? Was there a different `TypeError` with `keyword argument` that was confusing it?

Reply:
This was the failing test:

```
======================================================================
FAIL: test_populate_lenient_fallback (test_octodns_manager.TestManager)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/blanariu/work/cst/octodns/tests/test_octodns_manager.py", line 374, in test_populate_lenient_fallback
    manager._populate_and_plan('unit.tests.', [NoZone()], [])
AssertionError: TypeError not raised
-------------------- >> begin captured logging << --------------------
Manager: INFO: __init__: config_file=/Users/blanariu/work/cst/octodns/tests/config/simple.yaml
Manager: INFO: __init__:   max_workers=2
Manager: INFO: __init__:   include_meta=False
Manager: DEBUG: __init__:   configuring providers
YamlProvider[dump2]: DEBUG: __init__: id=dump2, directory=/var/folders/q5/fhmtlxb151d0jc5xlc66xhq00000gn/T/tmp2BMRQC, default_ttl=3600, enforce_order=1, populate_should_replace=0
YamlProvider[dump2]: DEBUG: __init__: id=dump2, apply_disabled=False, update_pcent_threshold=0.30, delete_pcent_threshold=0.30
YamlProvider[dump]: DEBUG: __init__: id=dump, directory=/var/folders/q5/fhmtlxb151d0jc5xlc66xhq00000gn/T/tmp2BMRQC, default_ttl=3600, enforce_order=1, populate_should_replace=0
YamlProvider[dump]: DEBUG: __init__: id=dump, apply_disabled=False, update_pcent_threshold=0.30, delete_pcent_threshold=0.30
YamlProvider[in]: DEBUG: __init__: id=in, directory=tests/config, default_ttl=3600, enforce_order=1, populate_should_replace=0
YamlProvider[in]: DEBUG: __init__: id=in, apply_disabled=False, update_pcent_threshold=0.30, delete_pcent_threshold=0.30
Manager: DEBUG: sync:   populating, zone=unit.tests., lenient=False
Manager: DEBUG: configured_sub_zones: subs=[u'subzone']
Zone: DEBUG: __init__: zone=Zone<unit.tests.>, sub_zones=set([u'subzone'])
Manager: WARNING: : provider NoLenient does not accept lenient param
Manager: DEBUG: sync:   planning, zone=unit.tests.
Manager: DEBUG: sync:   populating, zone=unit.tests., lenient=False
Manager: DEBUG: configured_sub_zones: subs=[u'subzone']
Zone: DEBUG: __init__: zone=Zone<unit.tests.>, sub_zones=set([u'subzone'])
Manager: WARNING: : provider NoZone does not accept lenient param
Manager: DEBUG: sync:   planning, zone=unit.tests.
--------------------- >> end captured logging << ---------------------
```

There is a difference between the error messages in the TypeError:
Python 2:
NoLenient: populate() got an unexpected keyword argument 'lenient'
NoZone: populate() got multiple values for keyword argument 'lenient'

Python 3:
NoLenient: populate() got an unexpected keyword argument 'lenient'
NoZone: populate() got multiple values for argument 'lenient'

Reply:
This one is interesting. That code has existed for a very long time, likely well before python 3 was even supported. 

What OS and version of python are you on?

```
coho:octodns ross$ uname -a
Darwin coho.lan 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64
coho:octodns ross$ python --version
Python 2.7.16
```

The multiple values for argument error message actually seems like something else is going wrong. I haven't been able to reproduce as of yet to dig in more.

Reply:
There's a difference between the error messages displayed. I haven't been able yet to find something related to how TypeError has changed.

Python 3.8.6:

```python
Python 3.8.6 (default, Dec 11 2020, 19:39:16) 
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
def f(x,y,z):
    return [x, y, z]
d = {'x': 1, 'y': 2, 'z': 3}
f(**d)
[1, 2, 3]
f(x=2, **d)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: f() got multiple values for keyword argument 'x'
f(2, **d)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: f() got multiple values for argument 'x'
```

Python 2.7.18:

```python
Python 2.7.18 (default, Jun 23 2021, 12:32:03) 
[GCC Apple LLVM 12.0.5 (clang-1205.0.22.9)] on darwin
def f(x,y,z):
    return [x,y,z]
d = {'x': 1, 'y': 2, 'z': 3}
f(**d)
[1, 2, 3]
f(x=2, **d)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: f() got multiple values for keyword argument 'x'
f(2, **d)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: f() got multiple values for keyword argument 'x'
```

Local setup:

```
(env) constantin:octodns blanariu$ uname -a
Darwin constantin.local 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64
(env) constantin:octodns blanariu$ python -V
Python 2.7.18
```

Reply:
Interesting. Thanks for digging further and documenting the findings.